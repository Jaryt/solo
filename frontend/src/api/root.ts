export default class RootAPI {
  protected static apiProtocol = process.env.REACT_APP_API_PROTOCOL || "https";
  protected static apiDomain = process.env.REACT_APP_API_DOMAIN;

  private static baseOptions: Partial<RequestInit> = {
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json"
    },
    redirect: "follow"
  };

  protected static async wrapFetch(url: string, options: Partial<RequestInit>) {
    const res = await fetch(
      `${this.apiProtocol}://${this.apiDomain}${url}`,
      options
    );
    if (!res.ok) {
      throw new Error(res.statusText);
    }
    return await res.json();
  }

  public static async get(
    url: string,
    options: Partial<RequestInit> | undefined = undefined
  ) {
    return this.wrapFetch(url, {
      ...options,
      ...this.baseOptions,
      method: "GET"
    });
  }

  public static async post(
    url: string,
    body: object,
    options: Partial<RequestInit> | undefined = undefined
  ) {
    return this.wrapFetch(url, {
      ...options,
      ...this.baseOptions,
      body: JSON.stringify(body),
      method: "POST"
    });
  }
}

export class AuthAPI extends RootAPI {
  protected static apiProtocol = process.env.REACT_APP_API_PROTOCOL || "https";
  protected static apiDomain = process.env.REACT_APP_AUTH_DOMAIN;
}
