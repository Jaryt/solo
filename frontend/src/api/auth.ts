import { AuthAPI } from "./root";

export default class AuthApi {
  public static login() {
    return AuthAPI.post("/login/", {});
  }
}
