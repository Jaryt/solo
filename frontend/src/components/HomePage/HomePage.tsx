import React, { useState, useCallback } from "react";
import LoginButton from "../LoginButton";
import * as api from "../../api";
import classes from "./HomePage.module.css";

const HomePage: React.FC = props => {
  const [data, setData] = useState("");

  const checkConnectivity = useCallback(async () => {
    setData("");
    try {
      await api.auth.apiConnectivityTest();
      setData("success");
    } catch (e) {
      setData(e.toString());
    }
  }, [setData]);

  return (
    <div className={classes.root}>
      <h1 className={classes.title}>SOLO Home Page</h1>
      <LoginButton onClick={checkConnectivity} />
      <div>{data}</div>
    </div>
  );
};

export default HomePage;
