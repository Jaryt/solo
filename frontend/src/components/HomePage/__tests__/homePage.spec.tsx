import React from "react";
import { render, fireEvent, wait } from "test-utils";
import HomePage from "../HomePage";
import { auth } from "../../../api";

jest.mock("../../../api/auth");

const authMocked = auth as jest.Mocked<typeof auth>;

describe("HomePage componenet", () => {
  beforeAll(() => {
    authMocked.login = jest.fn();
  });

  afterEach(() => {
    authMocked.login.mockReset();
  });

  afterAll(() => {
    authMocked.login.mockRestore();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<HomePage />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays success on api call success", async () => {
    authMocked.login.mockResolvedValue("test login return");
    const { getByText, queryByText } = render(<HomePage />);
    const loginButton = getByText("CAC Login");
    fireEvent.click(loginButton);
    await wait(() => {
      expect(queryByText("success")).toBeInTheDocument();
    });
  });

  it("displays error on api call failure", async () => {
    const err = new Error("some error message");
    authMocked.login.mockRejectedValue(err);
    const { getByText, queryByText } = render(<HomePage />);
    const loginButton = getByText("CAC Login");
    fireEvent.click(loginButton);
    await wait(() => {
      expect(queryByText(err.toString())).toBeInTheDocument();
    });
  });
});
