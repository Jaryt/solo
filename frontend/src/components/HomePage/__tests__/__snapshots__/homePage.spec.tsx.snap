// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage componenet matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <h1
      class="title"
    >
      SOLO Home Page
    </h1>
    <div
      class="root"
    >
      <button
        class="usa-button usa-button--big"
      >
        API Test
      </button>
    </div>
    <div />
  </div>
</DocumentFragment>
`;
